### PR TITLE
Add Karma, Jasmine, Webpack and Isparta ES6 coverage

### PR DESCRIPTION
**Pull request by nkbt**
Let me know if you have strong opinion against Jasmine and want to keep using tape, I'll figure out how to make them all work together. Jasmine seems to be the most common choice though.

Coverage looks pretty awesome now. Initial code is untouched.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -1,176 +1,176 @@
 {
```
Comment (ericelliott) [inline .eslintrc]: Why remove the `babel-eslint` parser, and why change the indentation?

File: karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+'use strict';
```
Comment (ericelliott) [inline karma.conf.js]: Why Karma?

File: package.json [hunk 3]
```diff
@@ -30,13 +28,21 @@
   "devDependencies": {
     "babel": "^5.5.8",
     "babel-eslint": "^3.1.15",
+    "babel-loader": "^5.1.4",
     "dependency-check": "^2.4.0",
     "eslint": "^0.23.0",
-    "faucet": "0.0.1",
-    "istanbul": "^0.3.15",
+    "isparta-loader": "^0.2.0",
+    "jasmine": "^2.3.1",
```
Comment (ericelliott) [inline package.json]: Absolutely not jasmine.

File: package.json [hunk 4]
```diff
@@ -30,13 +28,21 @@
...
+    "karma-coverage": "^0.4.2",
+    "karma-jasmine": "^0.3.5",
+    "karma-nyan-reporter": "0.0.60",
+    "karma-webpack": "^1.5.1",
+    "eslint": "^0.23.0",
     "nsp": "^1.0.1",
     "precommit-hook": "^2.0.1",
     "rimraf": "^2.4.0",
-    "tape": "^4.0.0"
+    "webpack": "^1.9.11"
```
Comment (ericelliott) [inline package.json]: Definitely willing to give this a try if it makes partial rebuilds and hot-patching live JS in Node and Chrome easier.

File: test/index-test.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+'use strict';
```
Comment (ericelliott) [inline test/index-test.js]: No.

File: test/webpack.tests.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+
+const testsContext = require.context('.', true, /\-test\.js$/);
```
Comment (ericelliott) [inline test/webpack.tests.js]: What is this?

**Comment by ericelliott** [comment]
## No Jasmine.

## Why Karma?

* Is it strictly necessary?
* Is there any way to get this working without it?
* If it's possible to make the coverage work without it, is that setup simpler, or more complex?

## Webpack

I'm totally willing to consider this as long as we're not doing crazy things with it that will make our codebase depend on it to work. I want to keep the door open to switch back if we find we need to later on.

File: .eslintrc (hunk 1 above)
Comment (nkbt) [inline .eslintrc]: 1. Firstly, it is not actually needed. ESLint by itself supports ES6 linting pretty well. Babel transformer slightly changes the code and some rules fail after those changes (I experienced it with my stricter linting config)
2. In the project there is 2 spaces indentation everywhere, so it is a little bit odd to see 4 spaces here (need .editorconfig #4)

File: karma.conf.js (hunk 2 above)
Comment (nkbt) [inline karma.conf.js]: Easy automation, works well with multiple browsers, code watch works well, good tools integration and support. Pretty common in the wild.

File: package.json (hunk 4 above)
Comment (nkbt) [inline package.json]: Yep, it is absolutely awesome. If you like, I can add webpack build with hot-reloading for the main code as well. I have couple of small npm components written like that - so good (like, https://github.com/nkbt/react-text-filter)

File: test/index-test.js (hunk 5 above)
Comment (nkbt) [inline test/index-test.js]: True, old good (not anymore) habit

File: test/webpack.tests.js (hunk 6 above)
Comment (nkbt) [inline test/webpack.tests.js]: Webpack stuff.
This way we have only one entry point instead of separate entry points for each test file. See https://github.com/webpack/karma-webpack#alternative-usage

Though if I think about it - for small modules we should keep it the default way. I switched to this alternative method on our main project because running tests (first run) on hundreds of test-files was extremely slow. Having one entry point fixed that in a very noticeable way.

File: package.json (hunk 3 above)
Comment (nkbt) [inline package.json]: No problems. I will put back tape then.

Comment (nkbt) [inline package.json]: Just for better understanding - what do you find bad in Jasmine?

I myself chose to use mocha + chai before, but then switched to jasmine, because it simply had more tooling around and used by all my peers, so it became easier to talk. So I have no strong opinion pro/against any of them.

File: .eslintrc (hunk 1 above)
Comment (nkbt) [inline .eslintrc]: I know it is a painful thing to check diffs with indentation changes on github =(. I did not touch a single rule (though, I would add some more ecmaFeatures, like in https://github.com/nkbt/esnext-quickstart/blob/master/src/.eslintrc)

Comment (ericelliott) [inline .eslintrc]: > Firstly, it is not actually needed.

It is if you want to use some of Babel's ES7 features, including the (commonly used) object rest and spread operators. I'd rather have access to features I use than have some strict rules I don't even use work properly.

> In the project there is 2 spaces indentation everywhere, so it is a little bit odd to see 4 spaces here (need .editorconfig #4)

This is a good point, but any time you make a change like indentation, it should be done in its own pull request so it's easy to see changes that have real meaning (such as rule changes, for instance).

Comment (nkbt) [inline .eslintrc]: I use them all on our main fairly large project and have no need in babel-eslint. Seriously.

Right, will separate indentation into another PR.

File: test/index-test.js (hunk 5 above)
Comment (nkbt) [inline test/index-test.js]: Hmm, `.eslintrc` should be a little bit updated, I reckon.

File: karma.conf.js (hunk 2 above)
Comment (ericelliott) [inline karma.conf.js]: I'm still listening and considering. Thanks for pushing me to learn more. So far these are all solved problems, with the tools I was already using, I just hadn't configured them, yet.

That said, if this is an easier solution, I'm open to it. Just glancing at the config though, it looked slightly more complicated and verbose than the npm scripts I planned to use. Kindof like Grunt but specifically for testing. My impression so far is that it solves problems we don't have.


## Watch

You can get the same watch features with webpack (which you also added) or watch, or [https://github.com/substack/watchify](watchify).


## Multiple browsers

The best way to do automated browser testing is with a cloud service provider such as [Sauce Labs](https://saucelabs.com) or [BrowserStack](https://www.browserstack.com/). Testling (the cli tool) is also a near-zero config solution for local browser testing that works perfectly with npm scripts.

## Integration and Support

How does it make life easier to integrate with tools like Jenkins, TravisCI, CircleCI, etc...? What does it give us that the npm scripts don't?

File: test/index-test.js (hunk 5 above)
Comment (nkbt) [inline test/index-test.js]: Oh yeah, this is why I can't find it in my code. It is deprecated rule.

http://eslint.org/docs/rules/global-strict.html

File: .eslintrc [hunk 7]
```diff
@@ -32,7 +30,6 @@
         "func-names": [0],
         "func-style": [0, "declaration"],
         "generator-star-spacing": [2, "after"],
-        "global-strict": [2, "always"],
```
Comment (nkbt) [inline .eslintrc]: 1. Depracated: http://eslint.org/docs/rules/global-strict.html
2. Does not work well with ES6 modules without `use strict`

File: package.json (hunk 3 above)
Comment (ericelliott) [inline package.json]: Almost all of [this applies to Jasmine](https://gist.github.com/ericelliott/d576f72441fc1b27dace#comment-1391854).

File: package.json (hunk 4 above)
Comment (ericelliott) [inline package.json]: That sounds great, but let's get this merged first so you can do that in a separate PR. =)

File: test/index-test.js (hunk 5 above)
Comment (ericelliott) [inline test/index-test.js]: Sorry, I wasn't talking about `'use strict';` - That's fine. I was talking about the entire file change. Please revert to the tape implementation. =)

File: test/webpack.tests.js (hunk 6 above)
Comment (ericelliott) [inline test/webpack.tests.js]: I'm not convinced that we need Karma messing with the test build at all. What's the advantage of doing that rather than using the same build tools and process we use for the rest of the repository assets?

I'm completely failing to see the logic of using Karma for the test build, except that karma also wants to bundle watch functionality that is probably better done outside the test config, too.. see my other comments about using npm scripts instead of Karma to automate all of that in a way that is consistent across the entire project, not just the tests.

Two different ways to configure the same things just seems like adding complexity with zero net gain.

File: karma.conf.js (hunk 2 above)
Comment (nkbt) [inline karma.conf.js]: Yes you can definitely do all these things manually or with bunch of other tools. It is just simpler to outsource all this work to something else. Like Karma. I can definitely explore other ways more (I did it before an it was not as simple as I would have thought), but I doubt it will be more expressive and less difficult in the result. Will almost certainly include some hacking with tools to mix and match them together.

I understand your concerns with locking yourself into some ecosystem, though find it reasonable in the case.

**Comment by ericelliott** [comment]
For all the ES6 module sources, we probably should stop including `'use strict';`, so even though that isn't what we meant, that change is good anyway. =)

File: .eslintrc (hunk 7 above)
Comment (ericelliott) [inline .eslintrc]: Cool. =)

**Comment by ericelliott** [comment]
Thanks a lot for all this great work. Sorry I'm pushing back so much on some of it. I just want to get this right and keep it as simple as possible. =)

File: package.json (hunk 3 above)
Comment (nkbt) [inline package.json]: Thanks. I tend to agree, but like BDD way and some nice features like mocking and spies. Will check what I can do with Tape for this project.

File: package.json (hunk 4 above)
Comment (nkbt) [inline package.json]: :+1:

File: .eslintrc (hunk 1 above)
Comment (ericelliott) [inline .eslintrc]: ES6 features:

No "classes" or "superInFunctions" -- everything else is great. :+1: 

[The Two Pillars of JavaScript](https://medium.com/javascript-scene/the-two-pillars-of-javascript-ee6f3281e7f3)
[How to Fix the ES6 Class Keyword](https://medium.com/javascript-scene/how-to-fix-the-es6-class-keyword-2d42bb3f4caf)
[Common Misconceptions About Inheritance in JavaScript](https://medium.com/javascript-scene/common-misconceptions-about-inheritance-in-javascript-d5d9bab29b0a)

**Comment by nkbt** [comment]
No worries. I am open to it. I am myself quite strict with the code and architecture when it comes to input from outside. Honestly did not expect this to be merged right away. More like start a discussion to explore the best.

File: .eslintrc (hunk 1 above)
Comment (nkbt) [inline .eslintrc]: Yeah, read them all and support 100%. I use them only because of React's new syntax for components.

**Comment by ericelliott** [comment]
Yeah. PRs will get merged faster if you split different things into different PRs. I would only group different things together if one change depended on another. =)

File: .eslintrc (hunk 1 above)
Comment (nkbt) [inline .eslintrc]: Created a separate PR for this and reverted the change.

File: karma.conf.js (hunk 2 above)
Comment (ericelliott) [inline karma.conf.js]: > It is just simpler to outsource all this work to something else. Like Karma.

To me it looks more complex, not simpler, but maybe I'm missing something. Can you show me how it simplifies config with some before and after examples?

File: package.json (hunk 3 above)
Comment (ericelliott) [inline package.json]: Sinnon works with just about everything (including tape).

Comment (nkbt) [inline package.json]: Yeah, I used to use SinonJS before. Sinon itself is pretty heavy one with heaps of functionality. Jasmine feels much lighter, and learning curve is not that steep.

File: karma.conf.js (hunk 2 above)
Comment (nkbt) [inline karma.conf.js]: I don't have a working example of this level of functionality not using Karma (or Mocha) at the moment. Will explore it and put together as a separate PR so we can compare which one looks better.

Comment (ericelliott) [inline karma.conf.js]: Great. =)

Comment (nkbt) [inline karma.conf.js]: Oops, checked docs https://github.com/douglasduteil/isparta

> Its intention is to be used with karma and karma-coverage, which provides code coverage reports using istanbul.

So far it is the only tool that gives a good es6 code coverage. I used isparta-loader, which just runs isparta instrumenter
```js
var instrumenter = new isparta.Instrumenter(assign(
        {
            embedSource: true,
            noAutoWrap: true
        },
        options
    ));
```

So it is possible to write some command-line utility to run isparta over the code and then run istanbul to get coverage report from it. But I am not sure it makes sense for the sake of not using karma. I would say getting pretty ES6 coverage is a good reason to use karma.

What do you reckon?

File: package.json (hunk 3 above)
Comment (nkbt) [inline package.json]: I'm having some issues with karma/tap, doesn't go as smooth as I expected. It's late here now, so will continue next time.

What I also found meanwhile, no matter which testing framework you use with karma, you can always use `karma-tap-reporter` to output report in a right format.

Comment (nkbt) [inline package.json]: what do you think about `testem`? Looks like pretty light testrunner, supports tap fine.

File: karma.conf.js (hunk 2 above)
Comment (ericelliott) [inline karma.conf.js]: Please clarify. Is this all the JavaScript we'd need to run isparta without karma / karma-coverage?

```js
var instrumenter = new isparta.Instrumenter(assign({
  embedSource: true,
  noAutoWrap: true
}, options));
```

**If so, that would be easy to do without the overhead and context switching of karma, and we should just use a little JS instead of a heavy-handed tool.**

**The overhead and context switching of karma may not be worth the cost of perfect-looking HTML coverage reports.**


## Why not Karma?

### We already have the mostly good-enough case

What we really need from a coverage automation tool is pass/fail and then then ability to figure out why it failed. That second part is the only awkward thing about the coverage reports without isparta.

**This problem *will* get fixed upstream in standard tools**


### The cost of Karma may be too high

**This repo will be the base of MANY other projects.** I don't think that getting perfect coverage reporting *right now* is worth the cost of introducing & **eventually extracting** Karma from every repo that inherits the config from this repo.

This last point may be true just for my personal needs, but consider that many other people may also use this repo, especially if I make it the basis for a scaffolding tool (a logical next step).

Comment (ericelliott) [inline karma.conf.js]: # RE Karma costs...

See [Developer Productivity Economics](https://twitter.com/_ericelliott/status/572924586775298048)

A context switch is similar to an interruption.

File: package.json (hunk 3 above)
Comment (ericelliott) [inline package.json]: Nope. Karma just got even more expensive. **Karma is vetoed**

Can you make this happen without Karma? If not, I'll close this PR.

**Comment by ericelliott** [comment]
# Karma is vetoed

I'm just gonna close this PR. If you can do this without Karma, please open a new PR that *only does the coverage* and doesn't change any of the other existing tech choices.


# Let's Try Webpack

I'm happy to consider webpack. Please open a separate PR for that change. If we find that it has higher costs than browserify, we can revert later.

**Comment by ericelliott** [comment]
If there were other improvements I agreed with (lint for example), let's make sure they're covered in separate PRs as well.

File: karma.conf.js (hunk 2 above)
Comment (nkbt) [inline karma.conf.js]: I think I know how to get coverage with isparta instrumenter, babel and istanbul. Will make another PR for that (hopefully today).

**Comment by nkbt** [comment]
Fair enough. This one goes off. I'll continue on getting coverage working and also make another one for webpack build

File: karma.conf.js (hunk 2 above)
Comment (ericelliott) [inline karma.conf.js]: You rock! \o/
:grinning:

**Comment by ericelliott** [comment]
:+1:
